src/Clan.ts

```diff
@@ -78,15 +80,44 @@ export class Clan {
       }
     }
 
-    const result = visitUrl(
-      `showclan.php?recruiter=1&whichclan=${clanId}&pwd&whichclan=${clanId}&action=joinclan&apply=Apply+to+this+Clan&confirm=on`
-    );
+    return Clan._join(clanId);
+  }
 
-    if (!result.includes("clanhalltop.gif")) {
-      throw new Error("Could not join clan");
+  /**
+   * Execute callback as a member of a clan
+   * and then restore prior membership
+   * @param clanIdOrName Clan id or name
+   */
+  static with<T>(clanIdOrName: string | number, callback: (clan: Clan) => T): T {
+    const startingClan = Clan.get();
+    const clan = Clan.join(clanIdOrName);
+    try {
+      return callback(clan);
+    } finally {
+      startingClan.join();
     }
+  }
 
-    return Clan.get();
+  /**
+   * Execute callback with items from a clan stash
+   * and then restore those items to the stash
+   * 
+   * During the execution of the callback, player will not be in the stash clan
+   * 
+   * @param clanIdOrName Clan id or name
+   */
+  static withStash<T>(clanIdOrName: string | number, items: Item[], callback: (borrowedItems: Item[]) => T): T {
+    const borrowed = Clan.with(clanIdOrName, clan => clan.take(items));
+    try {
+      return callback(borrowed);
+    } finally {
+      const returnedItems = Clan.with(clanIdOrName, clan => clan.put(borrowed));
```

Comment:
This should only change clans if we have taken stuff out - if nothing was available we shouldn't change.

Reply:
Great point!